{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Submitted by: \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# This fuction perfome data file handling (open and read the file contained data) when called\n",
    "# Parameter: file = the name of the file\n",
    "def read_file(file):\n",
    "    \n",
    "    # Statements to open the desired file\n",
    "    # Then, reading the opened file\n",
    "    with open(file) as f:\n",
    "        seq= f.read()       \n",
    "    \n",
    "    # Statement to terminate the function by returning the defined object by this function\n",
    "    return seq\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# This function create a directory to put the generated file from this script\n",
    "# Parameter: folder = the name of the folder that has to be created\n",
    "def directory(folder):\n",
    "    \n",
    "    # Statements to create a folder in the present location with the python module os\n",
    "    # That statements can check either the folder was created or not created.\n",
    "    # If the folder was not created, than create one with user defined folder name\n",
    "    # If the folder was created, than release a message. \n",
    "    import os                \n",
    "    if os.path.exists(folder)==True:      \n",
    "        template = \"Folder {} is already existed\".format(folder)\n",
    "        return template\n",
    "    else:                                    \n",
    "        dir_ = os.mkdir(folder)\n",
    "        return dir_\n",
    "\n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "# Calling the function directory(), so that the folder \"output\" will be created    \n",
    "directory(\"output\") \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function remove newline character from the string object\n",
    "# Parameter: file = the name of the file from which the newline has to be removed\n",
    "def remove_newline(file):\n",
    "    \n",
    "    \n",
    "    # Remove the newline character is worked by splitting the line on the basis of '\\n' and rejoining the splitted sequence\n",
    "    # Calling the function read_file() is needed to specify the file from which the newline character will be removed.\n",
    "    seq = read_file(file)\n",
    "    \n",
    "    \n",
    "    seq = seq.split('\\n')\n",
    "    \n",
    "    \n",
    "    seq = ''.join(i for i in seq)\n",
    "    \n",
    "    \n",
    "    return seq\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function extract exons from the gene sequence and create mRNA sequence\n",
    "# Parameters: file1 = gene sequence file-name, file2 = exon postions file-name\n",
    "\n",
    "def forming_mRNA(file1, file2):\n",
    "    \n",
    "    \n",
    "    # Statements remove newline character from the file1\n",
    "    # Read exons positions file and splitted into multiple line\n",
    "    seq = remove_newline(file1)\n",
    "    \n",
    "    exon_positions = read_file(file2)\n",
    "    \n",
    "    splitted_exon_positions = exon_positions.split(\"\\n\")\n",
    "    \n",
    "    \n",
    "    # Statements perform list comprehension to extract the exons from the gene sequence for the specific range of indices to a string object\n",
    "    # That specific range of indices are specified in the lines of exon positions file \n",
    "    # Each set of element is adding to the string in each step of the for loop\n",
    "    exon_specific_nucleotides = \"\"\n",
    "    for i in splitted_exon_positions:\n",
    "        start,end = i.split()\n",
    "        start,end = int(start),int(end)\n",
    "        exon_specific_nucleotides += seq[start:end]\n",
    "        \n",
    "    \n",
    "    # Statements to form the mRNA by replacing the letter \"T\" with \"U\"\n",
    "    mRNA = exon_specific_nucleotides\n",
    "    mRNA = mRNA.replace(\"T\",\"U\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    return mRNA\n",
    "        \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MELSWHVVFIALLSFSCWGSDWESDRNFISTAGPLTNDLLHNLSGLLGDQSSNFVAGDKDMYVCHQPLPTFLPEYFSSLH\\nASQITHYKVFLSWAQLLPAGSTQNPDEKTVQCYRRLLKALKTARLQPMVILHHQTLPASTLRRTEAFADLFADYATFAFH\\nSFGDLVGIWFTFSDLEEVIKELPHQESRASQLQTLSDAHRKAYEIYHESYAFQGGKLSVVLRAEDIPELLLEPPISALAQ\\nDTVDFLSLDLSYECQNEASLRQKLSKLQTIEPKVKVFIFNLKLPDCPSTMKNPASLLFSLFEAINKDQVLTIGFDINEFL\\nSCSSSSKKSMSCSLTGSLALQPDQQQDHETTDSSPASAYQRIWEAFANQSRAERDAFLQDTFPEGFLWGASTGAFNVEGG\\nWAEGGRGVSIWDPRRPLNTTEGQATLEVASDSYHKVASDVALLCGLRAQVYKFSISWSRIFPMGHGSSPSLPGVAYYNKL\\nIDRLQDAGIEPMATLFHWDLPQALQDHGGWQNESVVDAFLDYAAFCFSTFGDRVKLWVTFHEPWVMSYAGYGTGQHPPGI\\nSDPGVASFKVAHLVLKAHARTWHHYNSHHRPQQQGHVGIVLNSDWAEPLSPERPEDLRASERFLHFMLGWFAHPVFVDGD\\nYPATLRTQIQQMNRQCSHPVAQLPEFTEAEKQLLKGSADFLGLSHYTSRLISNAPQNTCIPSYDTIGGFSQHVNHVWPQT\\nSSSWIRVVPWGIRRLLQFVSLEYTRGKVPIYLAGNGMPIGESENLFDDSLRVDYFNQYINEVLKAIKEDSVDVRSYIARS\\nLIDGFEGPSGYSQRFGLHHVNFSDSSKSRTPRKSAYFFTSIIEKNGFLTKGAKRLLPPNTVNLPSKVRAFTFPSEVPSKA\\nKVVWEKFSSQPKFERDLFYHGTFRDDFLWGVSSSAYQIEGAWDADGKGPSIWDNFTHTPGSNVKDNATGDIACDSYHQLD\\nADLNMLRALKVKAYRFSISWSRIFPTGRNSSINSHGVDYYNRLINGLVASNIFPMVTLFHWDLPQALQDIGGWENPALID\\nLFDSYADFCFQTFGDRVKFWMTFNEPMYLAWLGYGSGEFPPGVKDPGWAPYRIAHAVIKAHARVYHTYDEKYRQEQKGVI\\nSLSLSTHWAEPKSPGVPRDVEAADRMLQFSLGWFAHPIFRNGDYPDTMKWKVGNRSELQHLATSRLPSFTEEEKRFIRAT\\nADVFCLNTYYSRIVQHKTPRLNPPSYEDDQEMAEEEDPSWPSTAMNRAAPWGTRRLLNWIKEEYGDIPIYITENGVGLTN\\nPNTEDTDRIFYHKTYINEALKAYRLDGIDLRGYVAWSLMDNFEWLNGYTVKFGLYHVDFNNTNRPRTARASARYYTEVIT\\nNNGMPLAREDEFLYGRFPEGFIWSAASAAYQIEGAWRADGKGLSIWDTFSHTPLRVENDAIGDVACDSYHKIAEDLVTLQ\\nNLGVSHYRFSISWSRILPDGTTRYINEAGLNYYVRLIDTLLAASIQPQVTIYHWDLPQTLQDVGGWENETIVQRFKEYAD\\nVLFQRLGDKVKFWITLNEPFVIAYQGYGYGTAAPGVSNRPGTAPYIVGHNLIKAHAEAWHLYNDVYRASQGGVISITISS\\nDWAEPRDPSNQEDVEAARRYVQFMGGWFAHPIFKNGDYNEVMKTRIRDRSLAAGLNKSRLPEFTESEKRRINGTYDFFGF\\nNHYTTVLAYNLNYATAISSFDADRGVASIADRSWPDSGSFWLKMTPFGFRRILNWLKEEYNDPPIYVTENGVSQREETDL\\nNDTARIYYLRTYINEALKAVQDKVDLRGYTVWSAMDNFEWATGFSERFGLHFVNYSDPSLPRIPKASAKFYASVVRCNGF\\nPDPATGPHACLHQPDAGPTISPVRQEEVQFLGLMLGTTEAQTALYVLFSLVLLGVCGLAFLSYKYCKRSKQGKTQRSQQE\\nLSPVSSF\\n'"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This function create protein from mRNA sequence\n",
    "# Parameters: file1 = the file-name of gene sequence, file2 = the file-name of exon-positions and\n",
    "# file3 = the filename of the code\n",
    "def creating_protein(file1 = \"gene_sequence.txt\",file2 = \"exon_positions.txt\", file3 = \"code.txt\"):\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Statements to call the functions, so that the the information of that functions can be used in this function\n",
    "    # Parameters are used as the arguments of the that functions\n",
    "    # mRNA sequence and code are defined by calling the functions\n",
    "    # The code is splitted into multiple lines \n",
    "    mRNA = forming_mRNA(file1,file2)\n",
    "    \n",
    "    code = read_file(file3)\n",
    "    code_ = code.split('\\n')\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Statements perform list comprehension to fillup a dictionary\n",
    "    # The keywords of the dictionary are the element of 1st column of the code (inside the code, each column is separated by the tab)\n",
    "    # Each set of elements is adding to the dictionary in each step of the for loop\n",
    "    cd = code_.copy()\n",
    "    \n",
    "    code_ = {}\n",
    "    \n",
    "    for i in cd:\n",
    "        j = i.split()\n",
    "        \n",
    "        code_[j[0]] = j[1]\n",
    "        \n",
    "    \n",
    "    \n",
    "    # Statements perform the list comprehension to form the string of amino acids\n",
    "    # Amino acid is formed by mapping every 3 nucleotides to a corresponded amino acid\n",
    "    # Each triplets of nucleotides in the form of mRNA is added with a string in each loop\n",
    "    # The loop started, when one start codon is found in the mRNA sequence\n",
    "    # The loop is stopped or breaken just before one triplets mapping into the amino acid 'X'\n",
    "    # 'X' is the amino acid symbol for end codon, thus loop breaks, when one of the three end codon is met\n",
    "    start_constant =  mRNA.find(\"AUG\")\n",
    "    length = int(len(mRNA[start_constant:])/3)\n",
    "    \n",
    "    protein = \"\"\n",
    "    for i in range(length):\n",
    "        start = start_constant+i*3\n",
    "        stop = start+3\n",
    "        value_code = code_[mRNA[start:stop]]\n",
    "        if value_code=='X':\n",
    "            break\n",
    "            \n",
    "        protein += value_code\n",
    "        \n",
    "\n",
    "\n",
    "    \n",
    "    # Statements perform list comprehension per 80 characters of the protein string\n",
    "    # A newline is created after each 80 characters of the protein string\n",
    "    # Each 80 characters and a newline character is added with a list at each loop\n",
    "    # When the elements of the list joined into single string, than line has created by the inserted newline character after each 80 characters\n",
    "    lines = []\n",
    "    for i in range(0,len(protein),80):\n",
    "        line = protein[i:i+80]+'\\n'\n",
    "        lines.append(line)\n",
    "    \n",
    "    \n",
    "    protein  = ''.join(i for i in lines) \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "        \n",
    "        \n",
    "\n",
    "    return protein\n",
    "\n",
    "\n",
    "# Calling the function creating_protein() to get the output of that function\n",
    "creating_protein()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function save the protein sequence as the text file format into the 'output' directory\n",
    "# Parameters: file1 = the file-name of gene sequence, file2 = the file-name of exon-positions and\n",
    "# file3 = the filename of the code\n",
    "def Output_protein_sequence(file1= \"gene_sequence.txt\",file2 = \"exon_positions.txt\", file3 = \"code.txt\"):\n",
    "    \n",
    "    \n",
    "    # Calling the functions, so that the information of that functions can be transferred into this function \n",
    "    protein = creating_protein(file1,file2,file3)\n",
    "    \n",
    "    directory(\"output\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Saving the protein sequence as the file format\n",
    "    # The folder 'output' was created and checked if the folder is existed, in the time of calling the function directory()\n",
    "    output = open(\"output/Protein_sequence.txt\", \"w\")\n",
    "    \n",
    "    output.write(protein)\n",
    "    \n",
    "    output.close()\n",
    "    \n",
    "\n",
    "# Calling this function is finally saving the protein sequence    \n",
    "Output_protein_sequence()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Name_of_the_disease': 'Congenital Lactase Deficiency',\n",
       " 'Normal_protein_length': 1927,\n",
       " 'Mutated_protein_length': 1389,\n",
       " 'Difference_normal_minus_mutated_protein_length': 538,\n",
       " 'Conclusion': \"(1) Protein has terminated upstream to the normal ending position due to the mutation and in the   place of termination, a stop codon has appeared. Thus, protein length has reduced and a 'Nonsense' mutation has   taken place. (2) Mutation hasn't changed the amino acid structure, unless the stop codon has appeared.\"}"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This function can determine the protein structure, length of protein and conclude if mutation happend in the position 30049, T>A\n",
    "# Parameters: file1 = the file-name of gene sequence, file2 = the file-name of exon-positions and\n",
    "# file3 = the filename of the code, position = the location where the point mutation happend in gene\n",
    "# point_mutation = the letter which has to be replaced by the original letter in the time of mutation\n",
    "def mutation(file1= \"gene_sequence.txt\",file2 = \"exon_positions.txt\", file3 = \"code.txt\", position = 30049, point_mutation = \"A\"):\n",
    "    \n",
    "    \n",
    "    # To create a mutated gene sequence, gene sequence file is read and the letter 'T' is replaced by the letter 'A'\n",
    "    # If letter 'T' isn't in the position of point-mutation, than this function will be terminated with an error message\n",
    "    gene_seq = remove_newline(file1)\n",
    "    \n",
    "    if gene_seq[position]==\"T\":\n",
    "        mutated_gene_seq = gene_seq[:position]+point_mutation+ gene_seq[position+1:]\n",
    "    else:\n",
    "        return \"Error, nucleotide must be T in the position\"\n",
    "        \n",
    "    # Mutated gene sequence is separated by the lines and each line is 80 character long    \n",
    "    lines = []\n",
    "    for i in range(0,len(mutated_gene_seq),80):\n",
    "        line = mutated_gene_seq[i:i+80]+'\\n'\n",
    "        lines.append(line)\n",
    "    \n",
    "    \n",
    "    mutated_gene_seq  = ''.join(i for i in lines)     \n",
    "        \n",
    "        \n",
    "    # Mutated gene is saving into the directory output\n",
    "    directory(\"output\")\n",
    "    mut_gene = open(\"output/Mutated_gene_sequence.txt\", \"w\")\n",
    "    \n",
    "    mut_gene.write(mutated_gene_seq)\n",
    "    \n",
    "    mut_gene.close()\n",
    "    \n",
    "    \n",
    "    # By the following statements, protein is created for the mutated gene while calling the function mutated_protein\n",
    "    # The filename of mutated protein sequence is passing as the argument of the function creating_protein, so that, protein is creating for the mutated gene\n",
    "    file = \"output/Mutated_gene_sequence.txt\"\n",
    "    \n",
    "    mutated_protein = creating_protein(file,file2,file3)\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Saving the mutated protein sequence\n",
    "    m_protein = open(\"output/Mutated_Protein_sequence.txt\", \"w\")\n",
    "    \n",
    "    m_protein.write(mutated_protein)\n",
    "    \n",
    "    m_protein.close()\n",
    "    \n",
    "    \n",
    "    # Read the normal protein and mutated protein file to compare the length between the normal and mutated protein \n",
    "    normal_protein = remove_newline(\"output/Protein_sequence.txt\")\n",
    "    \n",
    "    mutated_protein = remove_newline(\"output/Mutated_Protein_sequence.txt\")\n",
    "    \n",
    "    difference = len(normal_protein)-len(mutated_protein)\n",
    "    \n",
    "    \n",
    "    # Create the conclusion template which is depended on the value of difference and similarity or dissimalarity between the normal and mutated protein\n",
    "    if difference>0:\n",
    "        template1 = \"(1) Protein has terminated upstream to the normal ending position due to the mutation and in the   place of termination, a stop codon has appeared. Thus, protein length has reduced and a 'Nonsense' mutation has   taken place.\"\n",
    "    elif difference == 0:\n",
    "        template1 = \"(1) The length of mutated Protein is equal to the non-mutated protein.\"\n",
    "        \n",
    "        \n",
    "    if normal_protein[:len(mutated_protein)] == mutated_protein:\n",
    "        if difference>0:\n",
    "            template2 = \" (2) Mutation hasn't changed the amino acid structure, unless the stop codon has appeared.\"\n",
    "        else:\n",
    "            template2 = \" (2) Mutation hasn't changed the amino acid structure. Thus a 'Silent' mutation has happened\"\n",
    "    else:\n",
    "        template2 = \" (2) Mutation has changed the amino acid structure.\"\n",
    "        \n",
    "        \n",
    "    template = template1+template2    \n",
    "        \n",
    "        \n",
    "       \n",
    "    \n",
    "    # Putting the all statistics of the mutation and conclusion into a dictionary object\n",
    "    report = {\n",
    "              \"Name_of_the_disease\":\"Congenital Lactase Deficiency\",\n",
    "              \n",
    "              \"Normal_protein_length\": len(normal_protein),\n",
    "              \n",
    "              \"Mutated_protein_length\": len(mutated_protein),\n",
    "              \n",
    "              \"Difference_normal_minus_mutated_protein_length\": difference,\n",
    "              \n",
    "              \"Conclusion\": template\n",
    "             }\n",
    "    \n",
    "    # When function will be called, then that dictionary will be printed\n",
    "    return report\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Calling the mutation function\n",
    "mutation()\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
