{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this assignment, you will be working with FASTA files. It is a text-based format for \n",
    "representing either nucleic acid or protein sequences and in which nucleotides or amino \n",
    "acids are represented using single-letter codes.\n",
    "20 different standard amino acids exist in proteins (A, C, D, E, F, G, H, I, K, L, M, N, P, \n",
    "Q, R, S, T, V, W, Y), and these amino acids are divided into 4 categories according to \n",
    "their properties:\n",
    "\n",
    "\n",
    "\n",
    "- Positively charged: R, H, K\n",
    "- Negatively charged: D, E\n",
    "- Polar: N, C, Q, S, T, Y\n",
    "- Non-polar : A, I, L, M, F, P, W, V, G\n",
    "\n",
    "\n",
    "\n",
    "The input data needed for this assignment is provided in the attached archive uploaded \n",
    "on Blackboard. You can unzip this archive manually on your computer into a folder \n",
    "named \"FASTA_FILES\" or it can be done within your script using some specific library.\n",
    "The goal of this assignment is to ONLY use Python programming language to write a \n",
    "script that starts by displaying the following menu of the screen:\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "1) ALL\n",
    "2) Per category\n",
    "3) Within category\n",
    "4) Specific AA\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "1) If the user selects this option, the script will calculate the combined frequency\n",
    "and percentage of each of the 20 amino acid in the different FASTA files and \n",
    "print the results in a table sorted in reverse numerical order according to the \n",
    "percentage values. The table must contain 3 columns that correspond to the \n",
    "amino acid letter, its frequency and its percentage.\n",
    "\n",
    "\n",
    "\n",
    "2) If the user selects this option, the script will do the same as in (1), but the \n",
    "displayed results will be per amino acids category. The table will contain only 4 \n",
    "rows, each corresponding to an amino acid category.\n",
    "\n",
    "\n",
    "\n",
    "3) If the user selects this option, he/she will be prompted to choose one category\n",
    "from the four, and the script will display the frequency and percentage of each \n",
    "amino acid belonging to this category. In this case, the percentage of the amino \n",
    "acid is computed by dividing its frequency over the total number of amino acids \n",
    "belonging to the same category.\n",
    "\n",
    "\n",
    "4) If the user select this option, he/she will be prompted to enter the letter of the \n",
    "amino acid of interest, and the script will just compute and print the percentage \n",
    "of that specific amino acid.\n",
    "\n",
    "\n",
    "For the first three options (1, 2, and 3), the script will also return a plot to illustrate \n",
    "graphically the results of the frequency and percentage (2 y-axis scales can be used for \n",
    "each of the 2 metrics). You can use any data visualization library that is supported in \n",
    "Python Jupyter Notebooks (like Matplotlib, Seaborn, Plotly …). An interactive plot will \n",
    "be a plus. \n",
    "The first line of each FASTA file (starting with a > sign) is a description of the sequence\n",
    "and must not be considered in the calculations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Charbel younes\n",
    "# Python_Assignment_1\n",
    "# Import necassary python modules\n",
    "\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import ipympl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['.ipynb_checkpoints',\n",
       " 'P00451.fasta',\n",
       " 'P00533.fasta',\n",
       " 'P04637.fasta',\n",
       " 'P10275.fasta',\n",
       " 'P11362.fasta',\n",
       " 'P14373.fasta',\n",
       " 'P21802.fasta',\n",
       " 'P38398.fasta',\n",
       " 'P68871.fasta',\n",
       " 'Python_Assignment_1.ipynb',\n",
       " 'Q14524.fasta']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The jupyter notebook has opened in the FASTA_FILES. Now, we check the list of name in the FASTA_FILES\n",
    "Dir = os.listdir()\n",
    "Dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['P00451.fasta',\n",
       " 'P00533.fasta',\n",
       " 'P04637.fasta',\n",
       " 'P10275.fasta',\n",
       " 'P11362.fasta',\n",
       " 'P14373.fasta',\n",
       " 'P21802.fasta',\n",
       " 'P38398.fasta',\n",
       " 'P68871.fasta',\n",
       " 'Q14524.fasta']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove the name from the list that are not FASTA file\n",
    "for i in Dir:\n",
    "    if i[6:]!= '.fasta':\n",
    "        index = Dir.index(i)\n",
    "        Dir.pop(index)\n",
    "\n",
    "Dir         # Now we have a list object of the list of the name of FASTA files\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11055"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a list object in which, all the Amino acid are placed from all\n",
    "\n",
    "# The sequences founded in all FASTA files are combined into a single list object\n",
    "\n",
    "\n",
    "# Empty list object\n",
    "combined_seq = []\n",
    "\n",
    "\n",
    "\n",
    "# Appending the letter (Amino acid) in the combined_seq list from all the FASTA files\n",
    "for i in Dir:      # Looping through the FASTA files\n",
    "    for seq_record in SeqIO.parse(i,\"fasta\"):       # Loop through the each sequence record object of each FASTA file\n",
    "        \n",
    "        # Appending string format of amino acid symbol collected from the FASTA file as the new element of the list object\n",
    "        combined_seq.append(str(seq_record.seq))   # By the seq_record.seq method, only the sequence of protein is appending       \n",
    "\n",
    "\n",
    "        \n",
    "        \n",
    "        \n",
    "# Coonverting the list object into sequence object\n",
    "# The sequences founded in all FASTA files are combined into a single string object.\n",
    "combined_seq = \"\".join(i for i in combined_seq)     \n",
    "len(combined_seq)        # Length of the list object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a directory to save the created data\n",
    "\n",
    "\n",
    "if os.path.exists(\"Tables\") == False:   # Checking the if the directory is existed, if not existed, than creating a directory\n",
    "    dir1 = os.mkdir(\"Tables\")      #  a directory named \"Tables\" is created containing all the graphs representations\n",
    "\n",
    "if os.path.exists(\"Figures\") == False:    \n",
    "    dir2 = os.mkdir(\"Figures\")       #  a directory named \"Figures\" is created containg all the graph representations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# An inline to create interactive plot with matplotlib and ipympl in jupyter-notebook\n",
    "%matplotlib widget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A common code-template as the function is defining. so that, this template can be used with argument passing \n",
    "# in other functions\n",
    "# This code-template can perform plotting with appropriate argument\n",
    "\n",
    "def template_plotting(data,obj,filename):\n",
    "    \n",
    "    \"\"\"The function has three argument:\n",
    "       data = tabulated data in CSV file format,\n",
    "       obj = Amino acid or the name of Amino acid group\n",
    "       filename = the name of figure\"\"\"\n",
    "    \n",
    "    \n",
    "    # Defining the base figure in matplotlib\n",
    "    f, ax1 = plt.subplots()\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Defining the data\n",
    "    x = data['Name']\n",
    "    y = data['Frequency']\n",
    "    y_= data['Percentage']\n",
    "    \n",
    "    \n",
    "    \n",
    "    # All about the font of title or text is putting into dictionary object\n",
    "    font1 = {'family':'serif','color':'blue','size':20}\n",
    "    font2 = {'family':'serif','color':'red','size':15}\n",
    "    font3 = {'family':'serif','color':'blue','size':15}\n",
    "    font4 = {'family':'serif','color':'k','size':15}\n",
    "\n",
    "    \n",
    "    # color object\n",
    "    color = 'tab:red'\n",
    "    \n",
    "    \n",
    "    # Plotting the data\n",
    "    ax1.plot(x,y,'o', ms = 15,mfc = 'r',mec = 'r')\n",
    "    \n",
    "    \n",
    "    # Setting label for horizontal axis\n",
    "    ax1.set_xlabel(\"Amino-acid\", fontdict = font4)\n",
    "    \n",
    "    # Setting label for vertical axis in the left side\n",
    "    ax1.set_ylabel(\"Frequency\", fontdict = font2)\n",
    "    \n",
    "    # Coloring the scale-label of vertical axis in the left side\n",
    "    ax1.tick_params(axis='y', labelcolor=color)\n",
    "\n",
    "\n",
    "\n",
    "    color = 'tab:blue'\n",
    "    \n",
    "    # This property is coppying the axis of ax1, it is using to make two y-axis scale\n",
    "    ax2 = ax1.twinx()\n",
    "    \n",
    "    \n",
    "    ax2.plot(x,y_, ls = '--', c = 'b')\n",
    "    \n",
    "    \n",
    "    ax2.set_xlabel(\"Amino-acid\", fontdict = font4)\n",
    "    \n",
    "    # Setting label for vertical axis in the right side\n",
    "    ax2.set_ylabel(\"Percentage\", fontdict = font3)\n",
    "    \n",
    "    \n",
    "    # Coloring the scale-label of vertical axis in the right side\n",
    "    ax2.tick_params(axis='y', labelcolor=color)\n",
    "\n",
    "\n",
    "    # Template formatting to set the title, the parameter will be passed as the argument, when the function will be called\n",
    "    temp1 = \"Frequency and Percentage of \\n {obj}\".format(obj = obj)\n",
    "    \n",
    "    # Set the title with the template formatting object\n",
    "    plt.title(temp1, fontdict = font1)\n",
    "    \n",
    "    \n",
    "    # Tight layout resist the overlapping between two plots\n",
    "    f.tight_layout()\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Template formatting to set the filename, the parameter will be passed as the argument, when the function will be called   \n",
    "    temp2 =  \"Figures/{filename}.png\".format(filename = filename)\n",
    "    \n",
    "    # Saving fugure in the Figure directory\n",
    "    plt.savefig(temp2)\n",
    "    \n",
    "    # Output the plot\n",
    "    plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for option: All\n",
    "def All():\n",
    "    \n",
    "        \n",
    "    frequency = []        # Creating an empty list  for listing the frequency\n",
    "    percentage = []       # creating an empty list  for listing the percentage\n",
    "    \n",
    "    \n",
    "    # List of 20 Amino acid \n",
    "    li = (\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\")\n",
    "        \n",
    "    # Appending the frequency and percentage for all Amino acid\n",
    "    for i in li:          # Looping through the amino acid symbols list\n",
    "        frequency_ = combined_seq.count(i)    # Counting the frequency of each Amino acid from the combined sequence\n",
    "        \n",
    "        percentage_ = (combined_seq.count(i)/len(combined_seq))*100   # Counting the percentage of each Amino acid from the combined sequence\n",
    "        \n",
    "        frequency.append(frequency_)      # Updating the list  \n",
    "        \n",
    "        percentage.append(percentage_)\n",
    "            \n",
    "    \n",
    "    # Creating a dictionary with that three list: Name of Amino acid, frequency and percentage\n",
    "    dt = {\"Name\": li,\n",
    "          \"Frequency\": frequency,\n",
    "          \"Percentage\": percentage}\n",
    "\n",
    "    \n",
    "    # Creating a pandas data-frame object by the above dictionary\n",
    "    dtf = pd.DataFrame(dt)\n",
    "    \n",
    "    \n",
    "    # Sort values in reverse numerical order\n",
    "    dtf = dtf.sort_values(by='Percentage', ascending = False)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "        \n",
    "    \n",
    "    # Convert the dataframe into CSV file visible in the directory tables created above\n",
    "    table1 = dtf.to_csv(\"Tables/Table_1.CSV\", index = False)\n",
    "    \n",
    "    \n",
    "    # Read the CSV file\n",
    "    table1 = pd.read_csv(\"Tables/Table_1.CSV\")\n",
    "    \n",
    "    \n",
    "    # Print the CSV file\n",
    "    print(\"Table: All\",'\\n',table1, '\\n','\\n')\n",
    "    \n",
    "    \n",
    "    # Calling the function that can plot the generated data from this code block\n",
    "    template_plotting(table1,\"Amino acid\",\"All\")\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for option: All\n",
    "def Per_category():\n",
    "    frequency_p_c = []         # Creating an empty list  for listing the frequency\n",
    "    percentage_p_c = []        # creating an empty list  for listing the percentage\n",
    "    \n",
    "    \n",
    "    # List of Positively charged Amino acid group \n",
    "    p_c = [\"R\", \"H\", \"K\"]\n",
    "\n",
    "    # Appending the frequency and percentage of group of Amino acid\n",
    "    for i in p_c:       # Looping through the Amino acid symbols from the list of Amino acid group \n",
    "        frequency_ = combined_seq.count(i)      # Counting the  frequency of each Amino acid\n",
    "        percentage_ = (combined_seq.count(i)/len(combined_seq))*100    # Counting the percenatge of each Amino acid\n",
    "        \n",
    "        frequency_p_c.append(frequency_)      # Upadating the list \n",
    "        percentage_p_c.append(percentage_)\n",
    "        \n",
    "        \n",
    "    percentage_p_c = sum(percentage_p_c)      # Sum of percentage\n",
    "\n",
    "    frequency_p_c = sum(frequency_p_c)        # Sum of frequency\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    frequency_n_c = []\n",
    "    percentage_n_c = []\n",
    "\n",
    "    # List of Negatively charged Amino acid group\n",
    "    n_c = [\"D\",\"E\"]\n",
    "\n",
    "\n",
    "    for i in n_c:\n",
    "        frequency_ = combined_seq.count(i)\n",
    "        percentage_ = (combined_seq.count(i)/len(combined_seq))*100\n",
    "    \n",
    "        frequency_n_c.append(frequency_)\n",
    "        percentage_n_c.append(percentage_)\n",
    "        \n",
    "        \n",
    "    percentage_n_c = sum(percentage_n_c)\n",
    "\n",
    "    frequency_n_c = sum(frequency_n_c) \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    frequency_po = []\n",
    "    percentage_po = []\n",
    "\n",
    "    # List of Polar Amino acid group\n",
    "    po = [\"N\", \"C\", \"Q\", \"S\", \"T\", \"Y\"]\n",
    "\n",
    "\n",
    "    for i in po:\n",
    "        frequency_ = combined_seq.count(i)\n",
    "        percentage_ = (combined_seq.count(i)/len(combined_seq))*100\n",
    "    \n",
    "        frequency_po.append(frequency_)\n",
    "        percentage_po.append(percentage_)\n",
    "        \n",
    "        \n",
    "    percentage_po = sum(percentage_po)\n",
    "\n",
    "    frequency_po = sum(frequency_po) \n",
    "    \n",
    "    \n",
    "    \n",
    "    frequency_n_po = []\n",
    "    percentage_n_po = []\n",
    "\n",
    "    # List of Non-polar Amino acid group\n",
    "    n_po = [\"A\", \"I\", \"L\", \"M\", \"F\", \"P\", \"W\", \"V\", \"G\"]\n",
    "\n",
    "\n",
    "    for i in n_po:\n",
    "        frequency_ = combined_seq.count(i)\n",
    "        percentage_ = (combined_seq.count(i)/len(combined_seq))*100\n",
    "    \n",
    "        frequency_n_po.append(frequency_)\n",
    "        percentage_n_po.append(percentage_)\n",
    "        \n",
    "        \n",
    "    percentage_n_po = sum(percentage_n_po)\n",
    "\n",
    "    frequency_n_po = sum(frequency_n_po)  \n",
    "    \n",
    "    \n",
    "    # List of the name of the Amino acid group\n",
    "    Name = ['Positively Charged','Negatively Charged','Polar','Non-polar']\n",
    "    \n",
    "    # List of the frequency of the Amino acid group\n",
    "    frequency = [frequency_p_c,frequency_n_c,frequency_po,frequency_n_po]\n",
    "    \n",
    "    # List of the percentage of the Amino acid group\n",
    "    percentage = [percentage_p_c,percentage_n_c,percentage_po,percentage_n_po]\n",
    "\n",
    "    \n",
    "    # Creating a dictionary with that three list: Name of Amino acid, frequency and percentage\n",
    "    dic = {\"Name\": Name,\n",
    "          \"Frequency\": frequency,\n",
    "          \"Percentage\": percentage}\n",
    "\n",
    "    # Creating a pandas data-frame object by the above dictionary\n",
    "    dtf = pd.DataFrame(dic)\n",
    "    \n",
    "    # Sort values in reverse numerical order\n",
    "    dtf = dtf.sort_values(by='Percentage', ascending = False)\n",
    "    \n",
    "    \n",
    "   \n",
    "\n",
    "    # Creating the CSV file visible in the tables directory\n",
    "    table2 = dtf.to_csv(\"Tables/Table_2.CSV\", index = False)\n",
    "    \n",
    "    # Reading the CSV file\n",
    "    table2 = pd.read_csv(\"Tables/Table_2.CSV\")\n",
    "    \n",
    "    # printing the CSV file\n",
    "    print(\"Table: Per category\",'\\n',table2,'\\n','\\n')\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Calling the function that can plot the generated data by this script with proper argument \n",
    "    template_plotting(table2,\"categories of Amino acid\",\"Per_category\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"This code-template can be performed to calculate and to display the frequency and percentage of each\n",
    "amino acid belonging to this category with appropriate argument\"\"\"\n",
    "\n",
    "def template_within_category(li,filename,header):\n",
    "    \n",
    "    \"\"\"The function has three argument:\n",
    "       li = list of Amino acid group,\n",
    "       filename = the name of CSV file and figure,\n",
    "       header = The title for the output\"\"\"\n",
    "    \n",
    "    \n",
    "    frequency = []       # Creating an empty list  for listing the frequency     \n",
    "    percentage = []        # creating an empty list  for listing the percentage\n",
    "    \n",
    "   \n",
    "    # Appending the frequency and percentage of group of Amino acid\n",
    "    for i in li:      # Looping through the Amino acid symbols from the list of Amino acid group \n",
    "        frequency_ = combined_seq.count(i)       # Counting the  frequency of each Amino acid\n",
    "    \n",
    "        \n",
    "        \n",
    "        frequency.append(frequency_)     # Updating the list object \n",
    "        \n",
    "    \n",
    "    s_fre = sum(frequency)     # Sum of the frequency\n",
    "    \n",
    "    for i in frequency:        # Looping through the list frequency\n",
    "        p = (i/s_fre)*100        # Calculating percenatge of each Amino acid belonging to a category\n",
    "        percentage.append(p)       # Updating the list object\n",
    "        \n",
    "    \n",
    "    # Creating a dictionary with that three list: Name of Amino acid, frequency and percentage\n",
    "    dic = {\"Name\":li,\n",
    "           \"Frequency\":frequency,\n",
    "           \"Percentage\":percentage}\n",
    "\n",
    "    # Creating a pandas data-frame object by the above dictionary\n",
    "    dtf = pd.DataFrame(dic)\n",
    "    \n",
    "    # Sort values in reverse numerical order\n",
    "    dtf = dtf.sort_values(by='Percentage', ascending = False)\n",
    "    \n",
    "   \n",
    "    # Template formatting to set the filename, the parameter will be passed as the argument, when the function will call\n",
    "    temp1 = \"Tables/{filename}.CSV\".format(filename = filename)\n",
    "    \n",
    "    \n",
    "    # Creating the CSV file\n",
    "    table = dtf.to_csv(temp1, index = False)\n",
    "    \n",
    "    # Read the CSV file\n",
    "    table = pd.read_csv(temp1)\n",
    "    \n",
    "    \n",
    "    # Template formatting to set the title, the parameter will be passed as the argument, when the function will call\n",
    "    temp2 = \"Table: {header}\".format(header = header)\n",
    "    \n",
    "    # Printing the CSV file\n",
    "    print(temp2,'\\n',table,'\\n','\\n')\n",
    "    \n",
    "    # Template formatting to set the argument of another function\n",
    "    temp3 = \"{stc} Amino acid\".format(stc = header)\n",
    "    \n",
    "    # Calling the function that can plot the generated data by this script with proper argument \n",
    "    template_plotting(table,temp3,filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for sub-option: Positively charged\n",
    "def Positively_charged():\n",
    "    \n",
    "    # List of Positively charged Amino acid group \n",
    "    li = [\"R\", \"H\", \"K\"]\n",
    "    \n",
    "    # Filename of tabulated data that will be generate after function calling\n",
    "    filename = \"Positively_charged\"\n",
    "    \n",
    "    # Title of tabulated data that for printing\n",
    "    header = \"Positively charged\"\n",
    "    \n",
    "    # Calling the function that will print and plot the tabulated data of the Amino-acid group\n",
    "    template_within_category(li,filename,header)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Function for sub-option: Negatively charged    \n",
    "def Negatively_charged():    \n",
    "    \n",
    "    # List of Negatively charged Amino acid group \n",
    "    li = [\"D\",\"E\"]\n",
    "    \n",
    "    \n",
    "    filename = \"Negatively_charged\"\n",
    "    \n",
    "    header = \"Negatively charged\"\n",
    "    \n",
    "    template_within_category(li,filename,header)\n",
    "    \n",
    "    \n",
    "   \n",
    "    \n",
    "# Function for sub-option: Polar\n",
    "def Polar():\n",
    "    \n",
    "     # List of Polar Amino acid group \n",
    "    li = [\"N\", \"C\", \"Q\", \"S\", \"T\", \"Y\"]\n",
    "    \n",
    "    filename = \"Polar\"\n",
    "    \n",
    "    header = \"Polar\"\n",
    "    \n",
    "    template_within_category(li,filename,header)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "# Function for sub-option: Non-polar    \n",
    "def Non_polar(): \n",
    "    \n",
    "    # List of Non-polar Amino acid group \n",
    "    n_po = [\"A\", \"I\", \"L\", \"M\", \"F\", \"P\", \"W\", \"V\", \"G\"]\n",
    "    \n",
    "    filename = \"Non_polar\"\n",
    "    \n",
    "    header = \"Non-polar\"\n",
    "    \n",
    "    template_within_category(li,filename,header)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for option: Specific_AA\n",
    "\n",
    "def Specific_AA(letter):\n",
    "    \n",
    "    \n",
    "    percentage = []    # Creating an empty list object for listing the percentage \n",
    "   \n",
    "    # List of Amino acid\n",
    "    li = (\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\")\n",
    "    \n",
    "    # Dictionary object\n",
    "    # Name of Amino acid is the keyword\n",
    "    # Calculated percentage of Amino acid is the value\n",
    "    f_p = {i:(combined_seq.count(i)/len(combined_seq))*100 for i in li}\n",
    "    \n",
    "    \n",
    "    # Separating the percenatge of the Amino acid of specific later\n",
    "    # That letter will be apssed as the argument, when function will be called\n",
    "    item = f_p[letter]\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Template formatting to set the title, the parameter will be passed as the argument, when the function will call\n",
    "    template = \"Amino acid {letter}: Percentage{p}%\".format(letter = letter, p = item)\n",
    "    \n",
    "    \n",
    "    # Printing the percentage of the Amino acid of specific letter\n",
    "    print(\"Specific AA:\", '\\n', template,'\\n','\\n')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table: All \n",
      "    Name  Frequency  Percentage\n",
      "0     L       1009    9.127092\n",
      "1     S        947    8.566260\n",
      "2     E        784    7.091814\n",
      "3     P        653    5.906829\n",
      "4     G        651    5.888738\n",
      "5     A        618    5.590231\n",
      "6     V        618    5.590231\n",
      "7     T        614    5.554048\n",
      "8     K        600    5.427408\n",
      "9     D        522    4.721845\n",
      "10    R        507    4.586160\n",
      "11    I        490    4.432384\n",
      "12    N        487    4.405246\n",
      "13    Q        475    4.296698\n",
      "14    F        406    3.672546\n",
      "15    Y        308    2.786070\n",
      "16    M        281    2.541836\n",
      "17    C        258    2.333786\n",
      "18    H        258    2.333786\n",
      "19    W        140    1.266395 \n",
      " \n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6befb66d6d8b4b31b1f15b1c30fe5afb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table: Per category \n",
      "                  Name  Frequency  Percentage\n",
      "0           Non-polar       4866   44.016282\n",
      "1               Polar       3089   27.942108\n",
      "2  Positively Charged       1365   12.347354\n",
      "3  Negatively Charged       1306   11.813659 \n",
      " \n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "773816fbc2264035b514c930949d999f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table: Positively charged \n",
      "   Name  Frequency  Percentage\n",
      "0    K        600   43.956044\n",
      "1    R        507   37.142857\n",
      "2    H        258   18.901099 \n",
      " \n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "06870bb31f124398889aad4d63064f05",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table: Negatively charged \n",
      "   Name  Frequency  Percentage\n",
      "0    E        784   60.030628\n",
      "1    D        522   39.969372 \n",
      " \n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2723a3012a984f6f9c9a8a0e66716cb1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table: Polar \n",
      "   Name  Frequency  Percentage\n",
      "0    S        947   30.657171\n",
      "1    T        614   19.876983\n",
      "2    N        487   15.765620\n",
      "3    Q        475   15.377145\n",
      "4    Y        308    9.970864\n",
      "5    C        258    8.352218 \n",
      " \n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "44d49dba377e44b6928ae9d35d7f19a8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table: Non-polar \n",
      "    Name  Frequency  Percentage\n",
      "0     L       1009    9.495577\n",
      "1     S        947    8.912102\n",
      "2     E        784    7.378129\n",
      "3     P        653    6.145304\n",
      "4     G        651    6.126482\n",
      "5     A        618    5.815923\n",
      "6     V        618    5.815923\n",
      "7     T        614    5.778280\n",
      "8     K        600    5.646527\n",
      "9     D        522    4.912479\n",
      "10    R        507    4.771316\n",
      "11    I        490    4.611331\n",
      "12    N        487    4.583098\n",
      "13    Q        475    4.470168\n",
      "14    F        406    3.820817\n",
      "15    Y        308    2.898551\n",
      "16    M        281    2.644457\n",
      "17    C        258    2.428007\n",
      "18    H        258    2.428007\n",
      "19    W        140    1.317523 \n",
      " \n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d79f5531dc6642e29651a38e290e36c0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please, select a letter from Combobox and click again \n",
      " \n",
      "\n",
      "Specific AA: \n",
      " Amino acid A: Percentage5.59023066485753% \n",
      " \n",
      "\n",
      "Specific AA: \n",
      " Amino acid C: Percentage2.333785617367707% \n",
      " \n",
      "\n",
      "Specific AA: \n",
      " Amino acid Y: Percentage2.7860696517412937% \n",
      " \n",
      "\n",
      "Specific AA: \n",
      " Amino acid W: Percentage1.2663952962460425% \n",
      " \n",
      "\n",
      "Specific AA: \n",
      " Amino acid T: Percentage5.554047942107643% \n",
      " \n",
      "\n",
      "Specific AA: \n",
      " Amino acid S: Percentage8.56625961103573% \n",
      " \n",
      "\n",
      "Specific AA: \n",
      " Amino acid R: Percentage4.586160108548168% \n",
      " \n",
      "\n",
      "Specific AA: \n",
      " Amino acid Q: Percentage4.296698326549072% \n",
      " \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Import python tkinter package to let the user to interact with the \"options\", like a small menu\n",
    "import tkinter as tk\n",
    "from tkinter import *\n",
    "from tkinter import ttk\n",
    "\n",
    "\n",
    "# Defining the root tkinter object\n",
    "win = tk.Tk()\n",
    "\n",
    "# Background color of the GUI\n",
    "win.configure(background = 'skyblue')\n",
    "\n",
    "# Size of the GUI\n",
    "win.geometry('436x60')\n",
    "\n",
    "# Set the title of the GUI\n",
    "win.title(\"Assignment1: Python:  Seelect your option option\")\n",
    "\n",
    "# Defining the button for the option: All    \n",
    "button1 = ttk.Button(win, text = 'All', command = All)\n",
    "\n",
    "# Grid position of the button\n",
    "button1.grid(column = 1, row = 0)\n",
    "\n",
    "\n",
    "# Defining the button for the option: Per category\n",
    "button2 = ttk.Button(win, text = 'Per category', command = Per_category)\n",
    "button2.grid(column = 2, row = 0)\n",
    "\n",
    "\n",
    "\n",
    "# Defining the menubutton for the option: Within category\n",
    "menubutton1 = Menubutton(win, text=\"Within category\")\n",
    "menubutton1.grid(column = 3, row = 0 )\n",
    "\n",
    "# Create pull down menu\n",
    "menubutton1.menu = Menu(menubutton1, tearoff = 0)\n",
    "menubutton1[\"menu\"] = menubutton1.menu\n",
    "\n",
    "# Create the submenu button for the sub-option: Positively charged\n",
    "menubutton1.menu.add_command(label='Positively charged',command = Positively_charged )\n",
    "\n",
    "\n",
    "# Adding a separator between two submenu button\n",
    "menubutton1.menu.add_separator()\n",
    "\n",
    "# Create the submenu button for the sub-option: Negatively charged\n",
    "menubutton1.menu.add_command(label='Negatively charged', command = Negatively_charged)\n",
    "\n",
    "\n",
    "menubutton1.menu.add_separator()\n",
    "\n",
    "# Create the submenu button for the sub-option: Polar\n",
    "menubutton1.menu.add_command(label='Polar', command = Polar)\n",
    "\n",
    "\n",
    "menubutton1.menu.add_separator()\n",
    "\n",
    "#dropdown bar\n",
    "# Create the submenu button for the sub-option: Non-polar\n",
    "menubutton1.menu.add_command(label='Non-polar', command = Non_polar)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Function for the command of the button, to select the Amino acid from the Combobox\n",
    "def select():\n",
    "    letter = value.get()\n",
    "    if letter:\n",
    "        Specific_AA(letter)\n",
    "    else:\n",
    "        print(\"Please, select a letter from Combobox and click again\", '\\n','\\n')\n",
    "\n",
    "\n",
    "\n",
    "# The list of the name of the Amino acid\n",
    "li = (\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\")\n",
    "\n",
    "# Defining value for the combobox\n",
    "value = tk.StringVar()\n",
    "\n",
    "# Defining the combobox \n",
    "combo = ttk.Combobox(win, width = 12, textvariable = value)\n",
    "\n",
    "# Inserting the list of the name of Amino acid as the values of combobox\n",
    "combo['values'] = li\n",
    "\n",
    "combo.grid(column = 4, row = 1)\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "# Defining the button for the option: Specific AA\n",
    "button3 = ttk.Button(win, text = 'Specific AA', command = select)\n",
    "button3.grid(column = 4, row = 0)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Calling this function open the generated GUI \n",
    "win.mainloop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
